apps/homework/tasks.py

```diff
@@ -25,8 +27,42 @@ def __init__(self, *args, **kwargs):
         super(SubmissionPostException, self).__init__(*args, **kwargs)
 
 
-def retrieve_score_from_jupyter_notebook(fd, submission):
-    content = fd.readlines()
+def build_jupyter_notebook_submission(data_file, filename, submission):
+    numerator, denominator = retrieve_score_from_jupyter_notebook(data_file, filename, submission)
+    # This seek happens, because the retrieve_score_from_jupyter_notebook reads the file.
+    # The file position is reset here, so the next read, when the file is stored
+    # (the .save method below), all of the file contents are read.
+
+    data_file.seek(0)
+    submission.jupyter_notebook.save(filename, data_file)
+
+    submission_warnings = []
+    if denominator != submission.definition.jupyter_notebook_highest:
+        logger.warning(f'Score denominator is not equal to the one defined in homework definition. Warning in submission {submission.pk} for homework "{submission.definition.name}" (pk={submission.definition.pk}).')
+        submission_warnings.append('Score denominator is not equal to the one defined in homework definition.')
+    if numerator > submission.definition.jupyter_notebook_highest:
+        # Add warning to submission model
+        numerator = submission.definition.jupyter_notebook_highest
+        logger.warning(f'Score numerator greater than the maximum defined in homework definition. Warning in submission {submission.pk} for homework "{submission.definition.name}" (pk={submission.definition.pk}).')
+        submission_warnings.append('Score numerator greater than the maximum defined in homework definition.')
+    if numerator < submission.definition.jupyter_notebook_lowest:
+        # Add warning to submission model
+        logger.warning(f'Score numerator lower than the minimum defined in homework definition. Warning in submission {submission.pk} for homework "{submission.definition.name}" (pk={submission.definition.pk}).')
+        numerator = submission.definition.jupyter_notebook_lowest
+        submission_warnings.append('Score numerator lower than the minimum defined in homework definition.')
+
+    existing_messages = defaultdict(list, submission.reporting_messages)
+    existing_messages['warnings'].extend(submission_warnings)
+    submission.reporting_messages = existing_messages
+    submission.jupyter_score = numerator
+    submission.save()
+
+    if not submission.grades.first():
+        Grade.objects.create(submission=submission, jupyter_notebook_grade=submission.jupyter_score, evaluator=submission.definition.creator)
```

Comment:
`evaluator=submission.definition.creator`

is this correct?

Reply:
Yeah this is a non-nullable field that goes to an instructor. Currently there is only one instructor per klass, and that instructor created the definition to which this submission is being submitted. Hence, this is the shortest path to the instructor from the submission.

In the future, there may be multiple instructors per klass, so this will require some logic to figure out which one has been `assigned_to` this homework.